service/operator/operatorkit.go

```diff
@@ -0,0 +1,146 @@
+	// is already done at this point of the reconciliation loop.
+	ProcessCreateState(obj, createState interface{}) error
+	// ProcessDeleteState receives the custom object observed during TPR watches.
+	// It also receives the state intended to be deleted as provided by
+	// GetDeleteState. ProcessDeleteState only has to delete resources based on
+	// its provided input. All other reconciliation logic and state transformation
+	// is already done at this point of the reconciliation loop.
+	ProcessDeleteState(obj, deleteState interface{}) error
+}
+
+// ProcessCreate is a drop-in for an informer's AddFunc. It receives the custom
+// object observed during TPR watches and anything that implements Operator.
+// ProcessCreate takes care about all necessary reconciliation logic for create
+// events.
+//
+//     func addFunc(obj interface{}) {
+//         err := ProcessCreate(obj, operator)
+//         if err != nil {
+//             // error handling here
+//         }
+//     }
+//
+//     newResourceEventHandler := &cache.ResourceEventHandlerFuncs{
+//         AddFunc:    addFunc,
+//     }
+//
+func ProcessCreate(obj interface{}, operator Operator) error {
+	currentState, err := operator.GetCurrentState(obj)
+	if err != nil {
+		return microerror.MaskAny(err)
+	}
+
+	desiredState, err := operator.GetDesiredState(obj)
+	if err != nil {
+		return microerror.MaskAny(err)
+	}
+
+	createState, err := operator.GetCreateState(obj, currentState, desiredState)
+	if err != nil {
+		return microerror.MaskAny(err)
+	}
+
+	err = operator.ProcessCreateState(obj, createState)
+	if err != nil {
+		return microerror.MaskAny(err)
+	}
+
+	deleteState, err := operator.GetDeleteState(obj, currentState, desiredState)
+	if err != nil {
+		return microerror.MaskAny(err)
+	}
+
+	err = operator.ProcessDeleteState(obj, deleteState)
+	if err != nil {
+		return microerror.MaskAny(err)
+	}
+
+	return nil
+}
+
+// ProcessDelete is a drop-in for an informer's DeleteFunc. It receives the
+// custom object observed during TPR watches and anything that implements
+// Operator. ProcessDelete takes care about all necessary reconciliation logic
+// for delete events.
+//
+//     func deleteFunc(obj interface{}) {
+//         err := ProcessDelete(obj, operator)
+//         if err != nil {
+//             // error handling here
+//         }
+//     }
+//
+//     newResourceEventHandler := &cache.ResourceEventHandlerFuncs{
+//         DeleteFunc:    deleteFunc,
+//     }
+//
+func ProcessDelete(obj interface{}, operator Operator) error {
+	currentState, err := operator.GetCurrentState(obj)
+	if err != nil {
+		return microerror.MaskAny(err)
+	}
+
+	desiredState := operator.GetEmptyState()
+
+	deleteState, err := operator.GetDeleteState(obj, currentState, desiredState)
+	if err != nil {
+		return microerror.MaskAny(err)
+	}
+
+	err = operator.ProcessDeleteState(obj, deleteState)
+	if err != nil {
+		return microerror.MaskAny(err)
+	}
+
+	return nil
+}
```

Comment:
I went trough the code and I'm still not sure.

Say we have 10 resources to manage. Then the code looks more or less like:

```
func GetCurrentState() {
    // get state of resource 1
    // get state of resource 2
    // get state of resource 3
    // get state of resource 4
    // get state of resource 5
    // get state of resource 6
    // get state of resource 7
    // get state of resource 8
    // get state of resource 9
    // get state of resource 10
}

func GetDesiredState() {
    // create state of resource 1
    // create state of resource 2
    // create state of resource 3
    // create state of resource 4
    // create state of resource 5
    // create state of resource 6
    // create state of resource 7
    // create state of resource 8
    // create state of resource 9
    // create state of resource 10
}

func GetCreateState() {
    // make diff of current and desired state of resource 1
    // make diff of current and desired state of resource 2
    // make diff of current and desired state of resource 3
    // make diff of current and desired state of resource 4
    // make diff of current and desired state of resource 5
    // make diff of current and desired state of resource 6
    // make diff of current and desired state of resource 7
    // make diff of current and desired state of resource 8
    // make diff of current and desired state of resource 9
    // make diff of current and desired state of resource 10
}

func ProcessCreateState() {
    // apply diff of resource 1
    // apply diff of resource 2
    // apply diff of resource 3
    // apply diff of resource 4
    // apply diff of resource 5
    // apply diff of resource 6
    // apply diff of resource 7
    // apply diff of resource 8
    // apply diff of resource 9
    // apply diff of resource 10
}
```

There is a long "jump" between getting state of resource and applying diff.

I think following structure is easier to folow:

```
    // get state of resource 1
    // create desired state of resource 1
    // make diff of current and desired state of resource 1
    // apply diff of resource 1

    // get state of resource 2
    // create desired state of resource 2
    // make diff of current and desired state of resource 2
    // apply diff of resource 2

    ...
```

There is no enforcments on these steps, but this is how arguably our operators'
code should currently look like.

I'd like to know other reviewers opinion, as this is something we eventually
want to apply to **all** operators.